### PR TITLE
[feat] Add NamedPose

### LINKED ISSUE #1825: Add NamedPose API
## 🚀 Feature

base class api so that we we leverage 2d/3d pose estimation related problems for geometric computer vision.

- One of the common channels when describing a pose is describing the reference frames.
- We want the pose api to be compatible with optimization based pipelines like pytorch module.
- Compatible with other components such as [`Quaternion`](https://github.com/kornia/kornia/pull/1801), [`kornia.liegroups`](https://github.com/kornia/kornia/compare/liegroup-algebra) and opengl/opencv conversions https://github.com/kornia/kornia/pull/1823

## Initial proposal

```python

from kornia.core import Module
from kornia.geometry.liegroups import SE3d, SE2d

class NamedPose(Module):
    def __init__(
        self,
        a_pose_b: Union[SE3d, SE2d],
        frame_a: str,
        frame_b: str):
        ...

    # add conversion from/to
    def from_colmap(...):
       ...   

```

## Additional context

Links to consider: 

Based on community feedback in twitter

- http://paulfurgale.info/news/2014/6/9/representing-robot-pose-the-good-the-bad-and-the-ugly
- https://github.com/cvg/pixloc/blob/0ab0e795a443c67ccb948b6fa375393a5b98c093/pixloc/pixlib/geometry/wrappers.py#L103
- https://eigen.tuxfamily.org/dox/classEigen_1_1Rotation2D.html#a566207478cb88f9335c912d35913a0cd

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by cjpurackal**
#### Changes

Fixes #1825 

Opening this PR to get this going. Please suggest any functions/modifications that are required for this API. 

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/geometry/pose.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            raise ValueError(f"R must be either So2, So3, Quaternion, or Tensor, got {type(R)}")
+
+    def from_colmap():
+        raise NotImplementedError
+
+    def inverse(self) -> NamedPose:
+        return NamedPose(self.t_a_b.inverse(), self.frame_b, self.frame_a)
+
+    def transform(self, points: Tensor) -> Tensor:
+        return self.t_a_b * points
```
Comment (edgarriba) [inline kornia/geometry/pose.py:79]:
```suggestion
        return self.a_from_b * points_in_b
```
i 'd like that we follow same convention as in Sophus to describer transformations: https://strasdat.github.io/Sophus/latest/docs/Developer%20Guide/style-guide#rigid-body-transform-and-points-convention

File: kornia/geometry/pose.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            NamedPose(dst_from_src=rotation: Parameter containing:
+            tensor([1., -0., -0., -0.], requires_grad=True)
+            translation: x: 0.0
+            y: 0.0
+            z: 0.0,
+            frame_src: frame_b -> frame_dst: frame_a)
+        """
+        return NamedPose(self._dst_from_src.inverse(), self._frame_dst, self._frame_src)
+
+    def transform(self, points_in_src: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/geometry/pose.py:201]: we could also have a `compose` or similar to concatenate poses?

File: test/geometry/test_pose.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        pass
+
+    def test_mul(self, device, dtype):
+        b_from_a = NamedPose(
+            Se3.trans_x(torch.tensor([1.0], device=device, dtype=dtype)), frame_src="frame_a", frame_dst="frame_b"
+        )
+        c_from_b = NamedPose(
+            Se3.trans_y(torch.tensor([1.0], device=device, dtype=dtype)), frame_src="frame_b", frame_dst="frame_c"
+        )
+        c_from_a = b_from_a * c_from_b
```
Comment (edgarriba) [inline test/geometry/test_pose.py:43]:
```suggestion
        c_from_a = c_from_b * b_from_a
```

File: kornia/geometry/pose.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cjpurackal) [reply kornia/geometry/pose.py:201]: How will that differ from `__mul__`?

**Comment by cjpurackal** [comment]
@edgarriba anything more to be done here?

File: kornia/geometry/pose.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        cls,
+        rotation: So3 | So2 | Tensor | Quaternion,
+        translation: Tensor,
+        frame_src: str | None = None,
+        frame_dst: str | None = None,
+    ) -> NamedPose | None:
+        """Construct NamedPose from rotation and translation.
+
+        Args:
+            R : Rotation part of the pose.
```
Comment (edgarriba) [inline kornia/geometry/pose.py:115]:
```suggestion
            rotation: Rotation part of the pose.
```

File: kornia/geometry/pose.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+            T : Translation part of the pose.
```
Comment (edgarriba) [inline kornia/geometry/pose.py:116]:
```suggestion
            translation: Translation part of the pose.
```

File: kornia/geometry/pose.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            frame_src : Name of the source frame.
```
Comment (edgarriba) [inline kornia/geometry/pose.py:117]:
```suggestion
            frame_src: Name of the source frame.
```

File: kornia/geometry/pose.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+            frame_dst : Name of the destination frame.
```
Comment (edgarriba) [inline kornia/geometry/pose.py:118]:
```suggestion
            frame_dst: Name of the destination frame.
```

File: kornia/geometry/pose.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+        Returns:
+            NamedPose: NamedPose constructed from rotation and translation.
```
Comment (edgarriba) [inline kornia/geometry/pose.py:121]:
```suggestion
            NamedPose constructed from rotation and translation.
```

**Review by edgarriba — APPROVED** [review]
LGTM. Some docstrings typos, add also  to the docs. Maybe a warning stating that is an experimental api too?

**Review by johnnv1 — APPROVED** [review]
nice patch

**Comment by johnnv1** [comment]
@cjpurackal can you fix the typing here?

**Comment by cjpurackal** [comment]
@johnnv1 I think it should be good now.

**Review by johnnv1 — APPROVED** [review]

**Pull request merged**